File: backend/app/schemas/show.py
from datetime import datetime

from pydantic import BaseModel


class WebsiteInfo(BaseModel):
    id: int
    name: str
    display_name: str
    episode_count: int | None = None

    class Config:
        from_attributes = True


class ShowBase(BaseModel):
    title: str
    title_chinese: str | None = None
    poster_url: str | None = None
    description: str | None = None
    rating: float | None = None
    year: int | None = None
    status: str | None = None
    genres: list[str] | None = None
    total_episodes: int | None = None
    category: str | None = None


class ShowResponse(ShowBase):
    id: int
    slug: str | None = None
    created_at: datetime
    updated_at: datetime
    remote_updated_at: datetime | None = None
    websites: list[WebsiteInfo] = []

    class Config:
        from_attributes = True


class ShowListResponse(BaseModel):
    items: list[ShowResponse]
    total: int
    page: int
    page_size: int


class EpisodeResponse(BaseModel):
    id: int
    episode_number: int
    title: str | None = None
    external_url: str | None = None
    website_name: str | None = None

    class Config:
        from_attributes = True


class SourceResponse(BaseModel):
    id: int
    source_name: str
    source_key: str
    source_url: str | None = None
    source_type: str | None = None
    website_name: str | None = None

    class Config:
        from_attributes = True


class WatchProgressRequest(BaseModel):
    show_id: int
    episode_number: int
    progress_seconds: int
    duration_seconds: int | None = None
    episode_id: int | None = None


class WatchProgressResponse(BaseModel):
    id: int
    show_id: int
    episode_number: int
    progress_seconds: int
    duration_seconds: int | None = None
    completed: bool
    watched_at: datetime
    show: ShowBase | None = None

    class Config:
        from_attributes = True


class SyncResponse(BaseModel):
    shows: list[ShowResponse]
    watch_history: list[WatchProgressResponse]
    watchlist: list[int]  # show IDs
    timestamp: datetime


class HomeSection(BaseModel):
    title: str
    section_type: str  # continue_watching, recently_added, genre, completed
    shows: list[ShowResponse]


class HomeResponse(BaseModel):
    sections: list[HomeSection]

File: android-tv/app/src/main/java/com/donghuaflix/domain/model/Show.kt
package com.donghuaflix.domain.model

data class Show(
    val id: Int,
    val title: String,
    val titleChinese: String? = null,
    val slug: String? = null,
    val posterUrl: String? = null,
    val description: String? = null,
    val rating: Double? = null,
    val year: Int? = null,
    val status: String? = null,
    val genres: List<String> = emptyList(),
    val totalEpisodes: Int? = null,
    val category: String? = null,
    val websites: List<WebsiteInfo> = emptyList(),
    val watchProgress: WatchProgress? = null,
)

data class WebsiteInfo(
    val id: Int,
    val name: String,
    val displayName: String,
    val episodeCount: Int? = null,
)

data class Episode(
    val id: Int,
    val episodeNumber: Int,
    val title: String? = null,
    val externalUrl: String? = null,
    val websiteName: String? = null,
)

data class SubtitleTrack(
    val language: String,
    val label: String,
    val url: String,
)

data class VideoSource(
    val id: Int,
    val sourceName: String,
    val sourceKey: String,
    val sourceUrl: String? = null,
    val sourceType: String? = null,
    val websiteName: String? = null,
    val subtitles: List<SubtitleTrack> = emptyList(),
)

data class WatchProgress(
    val showId: Int,
    val episodeNumber: Int,
    val progressSeconds: Int = 0,
    val durationSeconds: Int? = null,
    val completed: Boolean = false,
)

data class StreamInfo(
    val url: String,
    val type: StreamType,
)

enum class StreamType {
    HLS, DASH, MP4, UNKNOWN;

    companion object {
        fun fromUrl(url: String): StreamType = when {
            url.contains(".m3u8") -> HLS
            url.contains(".mpd") -> DASH
            url.contains(".mp4") || url.contains(".mkv") || url.contains(".webm") -> MP4
            else -> UNKNOWN
        }
    }
}
